src/interpreter.jl

```diff
@@ -76,25 +85,19 @@ function Compiler.add_remark!(interp::CthulhuInterpreter, sv::InferenceState, ms
 end
 
 function Compiler.finish(state::InferenceState, interp::CthulhuInterpreter)
-    r = @invoke Compiler.finish(state::InferenceState, interp::AbstractInterpreter)
-    interp.unopt[Core.Compiler.any(state.result.overridden_by_const) ? state.result : state.linfo] = InferredSource(
+    res = @invoke Compiler.finish(state::InferenceState, interp::AbstractInterpreter)
+    key = Core.Compiler.any(state.result.overridden_by_const) ? state.result : state.linfo
+    interp.unopt[key] = InferredSource(
         copy(state.src),
         copy(state.stmt_info),
         isdefined(Core.Compiler, :Effects) ? state.ipo_effects : nothing,
         state.result.result)
-    return r
+    return res
 end
 
 function Compiler.transform_result_for_cache(interp::CthulhuInterpreter, linfo::MethodInstance,
         valid_worlds::WorldRange, @nospecialize(inferred_result))
-    if isa(inferred_result, OptimizationState)
-        opt = inferred_result
-        ir = opt.ir
-        if ir !== nothing
-            return OptimizedSource(ir, opt.src, opt.src.inlineable)
-        end
-    end
-    return inferred_result
+    return maybe_create_optsource(inferred_result, Effects())
```

Comment:
In https://github.com/JuliaLang/julia/pull/44660, I modified this function to pass in the effects. Maybe pull that part out of that PR and submit it separately, so we can use it here?